File: sdk/python/webgate.py
"""WebGate SDK for Python — discover and message website agents."""

import json
import urllib.request
from typing import Any, Optional


def discover(domain: str, scheme: str = "https", timeout: int = 5) -> dict[str, Any]:
    """Fetch WebGate manifest from a domain.

    Args:
        domain: e.g. "shop.com"
        scheme: "https" or "http"
        timeout: request timeout in seconds

    Returns:
        Parsed webgate.json as dict.

    Raises:
        ValueError: no webgate.json found or invalid format.
        urllib.error.URLError: network error.
    """
    url = f"{scheme}://{domain}/.well-known/webgate.json"
    req = urllib.request.Request(url, headers={"Accept": "application/json"})
    resp = urllib.request.urlopen(req, timeout=timeout)
    data = json.loads(resp.read().decode())
    if "endpoint" not in data:
        raise ValueError(f"Invalid webgate.json: missing 'endpoint' field")
    return data


def send(endpoint: str, message: dict[str, Any], timeout: int = 30) -> dict[str, Any]:
    """Send a message to a WebGate agent endpoint.

    Args:
        endpoint: agent's receiving URL (from discover())
        message: message payload, must include "from" and "content"
        timeout: request timeout in seconds

    Returns:
        Agent's response as dict.

    Raises:
        ValueError: missing required fields.
        urllib.error.URLError: network error.
    """
    if "from" not in message or "content" not in message:
        raise ValueError("Message must include 'from' and 'content' fields")
    data = json.dumps(message).encode()
    req = urllib.request.Request(
        endpoint,
        data=data,
        headers={"Content-Type": "application/json", "Accept": "application/json"},
    )
    resp = urllib.request.urlopen(req, timeout=timeout)
    return json.loads(resp.read().decode())


def visit(domain: str, intent: str, from_id: str = "anonymous", scheme: str = "https", **context) -> dict[str, Any]:
    """Discover a WebGate agent and send a first message. One-liner entry point.

    Args:
        domain: website domain to visit
        intent: what the visiting agent wants to do
        from_id: visiting agent's identity
        scheme: "https" or "http"
        **context: additional fields merged into the message

    Returns:
        Agent's response as dict.
    """
    manifest = discover(domain, scheme=scheme)
    message = {"from": from_id, "content": intent, **context}
    return send(manifest["endpoint"], message)

File: sdk/ts/webgate.ts
/**
 * WebGate SDK for TypeScript — discover and message website agents.
 */

export interface WebGateManifest {
  version: string;
  endpoint: string;
  [key: string]: unknown;
}

export interface WebGateMessage {
  from: string;
  content: string | object;
  [key: string]: unknown;
}

/**
 * Fetch WebGate manifest from a domain.
 */
export async function discover(domain: string, scheme = "https"): Promise<WebGateManifest> {
  const url = `${scheme}://${domain}/.well-known/webgate.json`;
  const resp = await fetch(url, {
    headers: { Accept: "application/json" },
  });
  if (!resp.ok) {
    throw new Error(`WebGate manifest not found at ${url} (HTTP ${resp.status})`);
  }
  const data: WebGateManifest = await resp.json();
  if (!data.endpoint) {
    throw new Error(`Invalid webgate.json: missing 'endpoint' field`);
  }
  return data;
}

/**
 * Send a message to a WebGate agent endpoint.
 */
export async function send(
  endpoint: string,
  message: WebGateMessage,
  options?: RequestInit,
): Promise<unknown> {
  if (!message.from || !message.content) {
    throw new Error("Message must include 'from' and 'content' fields");
  }
  const resp = await fetch(endpoint, {
    method: "POST",
    headers: { "Content-Type": "application/json", Accept: "application/json" },
    body: JSON.stringify(message),
    ...options,
  });
  return resp.json();
}

/**
 * Discover a WebGate agent and send a first message. One-liner entry point.
 */
export async function visit(
  domain: string,
  intent: string,
  fromId = "anonymous",
  context?: Record<string, unknown>,
): Promise<unknown> {
  const manifest = await discover(domain);
  const message: WebGateMessage = {
    from: fromId,
    content: intent,
    ...context,
  };
  return send(manifest.endpoint, message);
}
